[Offload] Initialize all platforms before plugin device creation

pbalcer opened the pull request:
Devices store a raw pointer to back to their owning Platform. Platforms are stored directly inside of a vector. Modifying this vector risks invalidating all the platform pointers stored in devices.

This change moves the host platform emplace to happen before device object creation, eliminating the risk of vector reallocation after platform pointers are set.

This popped up when trying to use https://github.com/llvm/llvm-project/pull/158900 with liboffload. The platforms vector was reallocated and liboffload segfaulted when iterating over the platforms. This is an attempt at the simplest possible fix.

Since the number of platforms is fixed, I initially wanted to refactor this code so that it became impossible to modify the platforms container after Context is created, but, when that patch grew large, I pivoted to this instead. Another option is to use shared_ptr/weak_ptr to allocate the Platform instead of storing it in the vector directly.

Given, https://github.com/llvm/llvm-project/issues/159636, which requires a non-trivial change to this same code anyway, I'm thinking I can do follow-up refactor that tackles both issues at the same time.
Thoughts?

jhuber6 reviewed (commented):
How difficult would it be to just make `SmallVector<ol_platform_impl_t, 4> Platforms{};` take a unique pointer instead? In the future I really want to remove this restriction that *all* platforms and devices must be initialized at init time, as it's extremely expensive if you don't end up needing them.

pbalcer committed:
[Offload] Initialize all platforms before plugin device creation

Devices store a raw pointer to back to their owning Platform.
Platforms are stored directly inside of a vector. Modifying this
vector risks invalidating all the platform pointers stored in devices.

This change moves the host platform emplace to happen before
device object creation, eliminating the risk of vector reallocation
after platform pointers are set.
